azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,342 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from knack.log import get_logger
+
+from azext_iot.common.shared import DeviceAuthApiType, DeviceAuthType
+from azext_iot.iothub.providers.base import IoTHubProvider
+from azext_iot.operations.hub import _iot_device_show, _iot_device_twin_show, _iot_device_module_twin_show, \
+    _iot_device_module_show
+from azext_iot.operations.hub import iot_device_list, iot_device_create, iot_device_delete, iot_device_twin_replace, \
+    iot_hub_configuration_create, iot_hub_configuration_list, iot_hub_configuration_delete, iot_device_module_list, \
+    iot_device_module_create, iot_device_module_twin_replace
+
+import json
+from tqdm import tqdm
+from typing import Optional
+
+logger = get_logger(__name__)
+
+
+class StateProvider(IoTHubProvider):
+    def __init__(
+        self,
+        cmd,
+        hub: Optional[str] = None,
+        rg: Optional[str] = None,
+        login: Optional[str] = None,
+        auth_type_dataplane: Optional[str] = None,
+    ):
+        super(StateProvider, self).__init__(
+            cmd=cmd,
+            hub_name=hub,
+            rg=rg,
+            login=login,
+            auth_type_dataplane=auth_type_dataplane
+        )
+        self.auth_type = auth_type_dataplane
+
+        # the login argument doesn't work later if auth type = "login" and a connection string isn't given initially
+        if login or auth_type_dataplane == "key":
+            self.login = self.target["cs"]
+        else:
+            self.login = None
+
+    def save_state(self, filename: str):
+        '''
+        Writes all hub configurations, device identities and device twins from the origin hub to a json file
+        '''
+
+        configs = iot_hub_configuration_list(cmd=self.cmd, hub_name=self.hub_name, resource_group_name=self.rg,
+                                             login=self.login, auth_type_dataplane=self.auth_type)
+
+        identities = iot_device_list(cmd=self.cmd, hub_name=self.hub_name, top=-1, resource_group_name=self.rg,
+                                     login=self.login, auth_type_dataplane=self.auth_type)
+
+        with open(filename, 'w', encoding='utf-8') as f:
+
+            json.dump(configs, f)
+            f.write('\n')
+
+            for i in tqdm(range(len(identities)), desc="Exporting devices"):
```

Comment:
include tqdm for config and identities - you can just have different instances (so it would be one tqdm for all the configs, followed by another for identities, and so on)

Reply:
even though the configs are all just being dumped at once? there's no iteration happening there